### PR TITLE
Try to allow memoization of components with state

### PR DESCRIPTION
**Pull request by darknoon**
Work in progress, via discussions with @adamjernst. Need to clean up the `id scopeIdentifier` parameter, which shouldn't be necessary.

### CODE REVIEW

**Comment by darknoon** [comment]
Related to #316.

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 1]
```diff
@@ -14,11 +14,41 @@
...
 #import "CKMacros.h"
 #import "CKInternalHelpers.h"
+#import "CKThreadLocalComponentScope.h"
+#import "CKComponentScopeHandle.h"
 
 #include <map>
 
 static NSString *CKComponentMemoizerThreadKey = @"CKComponentMemoizer";
 
+static bool CKCanUseMemoizedComponent(CKComponent *c, id scopeFrameIdentifier) {
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: Instead, use `c.scopeHandle.rootIdentifier` which is a unique identifier for its scope frame that persists across state updates.

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 2]
```diff
@@ -14,11 +14,41 @@
...
+  NSCAssert(s != nullptr, @"No thread local scope!");
+
+  if (!s) {
+    return false;
+  }
+
+  auto stackTop = s->stack.top();
+
+  // Look for the equivalent child frame
+  CKComponentScopeFrame *oldScopeFrame = [stackTop.equivalentPreviousFrame scopeFrameForClass:c.class identifier:scopeFrameIdentifier];
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: You can just make this `scopeFrameForHandleWithGlobalIdentifier:` then. Don't even need to pass in `c.class`.

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 3]
```diff
@@ -88,4 +88,29 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
   return self;
 }
 
+- (CKComponentScopeFrame *)scopeFrameForClass:(Class)componentClass identifier:(id)identifier;
+{
+  auto it = _children.find({componentClass, identifier});
```
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: This is going to be relatively less efficient when you're doing it based on global identifiers instead of class/identifiers; now it'll be an O(N) search instead of O(1). But given that the number of children will typically be ~5 at most, that should be just fine.

Comment (darknoon) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: Updating the diff now; slightly concerned about complexity but it seems to work for my use case at the moment.

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 4]
```diff
@@ -14,11 +14,47 @@
...
 
 #include <map>
 
+
 static NSString *CKComponentMemoizerThreadKey = @"CKComponentMemoizer";
 
+static bool CKCanUseMemoizedComponent(CKComponent *c) {
+  CKThreadLocalComponentScope *s = CKThreadLocalComponentScope::currentScope();
+
+  NSCAssert(s != nullptr, @"No thread local scope!");
```
Comment (darknoon) [inline ComponentKit/Core/CKComponentMemoizer.mm]: I should use CKAssert or whatever here

File: ComponentKit/Core/Scope/CKComponentScopeFrame.h [hunk 5]
```diff
@@ -34,4 +34,16 @@ struct CKComponentScopeFramePair {
 
 @property (nonatomic, strong, readonly) CKComponentScopeHandle *handle;
 
+struct _CKChildScopeFrameRet {
```
Comment (darknoon) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.h]: This is somewhat inelegant, but I need to return more than just one of these items… thoughts?

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 6]
```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
   return self;
 }
 
+- (void)injectScopeHierarchy:(_CKChildScopeFrameRet)oldFrame
+{
+  Class componentClass = oldFrame.cls;
+  id identifier = oldFrame.identifier;
+  CKAssert([oldFrame.frame isKindOfClass:[CKComponentScopeFrame class]], @"Can't inject with incorrect frame class.");
+
+  auto result = _children.insert({{componentClass, identifier}, oldFrame.frame});
```
Comment (darknoon) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: Please review this carefully, I'm not familiar enough with the code to know it it's really safe.

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 7]
```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
...
+    if (pr.second->_handle.globalIdentifier == globalIdentifier) {
+      return _CKChildScopeFrameRet{pr.first.componentClass, pr.first.identifier, pr.second};
+    }
+  }
+  return {nil, nil, 0};
+}
+
+- (BOOL)anyDescendentHasPendingUpdates:(const CKComponentStateUpdateMap &)stateUpdates
+{
+  auto it = stateUpdates.find(_handle.globalIdentifier);
```
Comment (darknoon) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: Is there a way we could use immutable ptrs to avoid this recursion in the no-change case, which should be very common?

**Comment by adamjernst** [comment]
Reviewing now...

File: ComponentKit/Core/CKComponent.mm [hunk 8]
```diff
@@ -300,4 +300,9 @@ - (id)scopeFrameToken
   return _scopeHandle ? @(_scopeHandle.globalIdentifier) : nil;
 }
 
+- (CKComponentScopeHandleIdentifier)globalIdentifier
+{
+  return _scopeHandle.globalIdentifier;
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponent.mm]: This already exists as `scopeFrameToken`, can we just use that?

File: ComponentKit/Core/CKComponentMemoizer.mm (hunk 4 above)
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: Yup

File: ComponentKit/Core/CKComponent.mm (hunk 8 above)
Comment (adamjernst) [inline ComponentKit/Core/CKComponent.mm]: Ah I see why you didn't and exposed this instead. Will consider...

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm (hunk 7 above)
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: Not as far as I can see because `stateUpdates` is just a flat map of ids -> updates.

We could annotate each node in the state tree with a hash-set of all the IDs below it, which would make this an O(1) operation here, but given the typical size of a state tree (10-15 elements) I don't think it's worth avoiding the recursion. For now let's do this, and come back to it if it's a hotspot (doubt it)

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 9]
```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
...
+  if (it != stateUpdates.end()) {
+    return YES;
+  }
+  // Look in children
+  for (auto &pr : _children) {
```
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: const auto &pr

will avoid copying, not that it matters much since they're just pairs

Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: (the copy is necessary since `stateUpdates` is const, so to give you a non-const pr it must copy the pair)

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 10]
```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
...
+{
+  for (auto &pr : _children) {
+    if (pr.second->_handle.globalIdentifier == globalIdentifier) {
+      return _CKChildScopeFrameRet{pr.first.componentClass, pr.first.identifier, pr.second};
+    }
+  }
+  return {nil, nil, 0};
+}
+
+- (BOOL)anyDescendentHasPendingUpdates:(const CKComponentStateUpdateMap &)stateUpdates
```
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: descendant not descendent

File: ComponentKit/Core/Scope/CKComponentScopeFrame.h (hunk 5 above)
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.h]: I think it's fine.

- Move it to the top of the file
- Rename to `CKChildScopeHierarchy` (drop underscore, lose abbreviation, and I like your term "scope hierarchy" below better than "child scope frame")

File: ComponentKit/Core/Scope/CKComponentScopeFrame.h [hunk 11]
```diff
@@ -34,4 +34,16 @@ struct CKComponentScopeFramePair {
...
+  Class cls;
+  id identifier;
+  CKComponentScopeFrame *frame;
+};
+
+- (_CKChildScopeFrameRet)childScopeFrameForOldIdentifier:(CKComponentScopeHandleIdentifier)globalIdentifier;
```
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.h]: `childScopeHierarchyForScopeFrameToken:(id)scopeFrameToken`

Inside the implementation just check if the scope frame token is an NSNumber and unwrap it if so. Kinda sucks as it's a hidden coupling with CKComponent's `scopeFrameToken`, but it is better than exposing another method on CKComponent (even if it's internal-only)

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 12]
```diff
@@ -14,11 +14,47 @@
...
 #import "CKInternalHelpers.h"
+#import "CKThreadLocalComponentScope.h"
+#import "CKComponentScopeHandle.h"
 
 #include <map>
 
+
 static NSString *CKComponentMemoizerThreadKey = @"CKComponentMemoizer";
 
+static bool CKCanUseMemoizedComponent(CKComponent *c) {
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: nit: { on new line

nit: don't need "CK" prefix if it's static

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 13]
```diff
@@ -14,11 +14,47 @@
...
+  NSCAssert(s != nullptr, @"No thread local scope!");
+  if (!s) {
+    return false;
+  }
+
+  auto stackTop = s->stack.top();
+  CKComponentScopeHandleIdentifier oldIdent = c.globalIdentifier;
+
+  // Look for the equivalent child frame
+  auto oldFrameRet = [stackTop.equivalentPreviousFrame childScopeFrameForOldIdentifier:oldIdent];
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: might as well `const auto`

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 14]
```diff
@@ -14,11 +14,47 @@
...
+  CKComponentScopeFrame *oldScopeFrame = oldFrameRet.frame;
+
+  // Make sure none of the descendents of the oldScopeFrame have pending state updates
+  if (oldScopeFrame && s->stateUpdates.size() > 0) {
+    if ([oldScopeFrame anyDescendentHasPendingUpdates:s->stateUpdates]) {
+      //NSLog(@"Cannot reuse component %@ because state has been updated.", c);
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: clean up these logs of course…

File: ComponentKit/Core/CKComponentMemoizer.mm [hunk 15]
```diff
@@ -149,6 +185,11 @@ id CKMemoize(CKMemoizationKey memoizationKey, id (^block)(void))
 {
   _CKComponentMemoizerImpl *impl = [_CKComponentMemoizerImpl currentMemoizer];
   CKComponent *component = [impl dequeueComponentForKey:memoizationKey];
+  if (component) {
+    if (!CKCanUseMemoizedComponent(component)) {
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentMemoizer.mm]: can we just move this check inside `dequeue`? Or do you prefer to keep it here for cleanliness of the queue? your call

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm (hunk 6 above)
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: Yup it's safe. You're injecting a scope frame from the *old* tree, which by now is immutable.

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 16]
```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
...
 
+- (void)injectScopeHierarchy:(_CKChildScopeFrameRet)oldFrame
+{
+  Class componentClass = oldFrame.cls;
+  id identifier = oldFrame.identifier;
+  CKAssert([oldFrame.frame isKindOfClass:[CKComponentScopeFrame class]], @"Can't inject with incorrect frame class.");
+
+  auto result = _children.insert({{componentClass, identifier}, oldFrame.frame});
+
+  CKAssert(result.second, @"Scope collision: attempting to create duplicate scope %@:%@", componentClass, identifier);
```
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: yup, good.

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 17]
```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
   return self;
 }
 
+- (void)injectScopeHierarchy:(_CKChildScopeFrameRet)oldFrame
+{
+  Class componentClass = oldFrame.cls;
+  id identifier = oldFrame.identifier;
+  CKAssert([oldFrame.frame isKindOfClass:[CKComponentScopeFrame class]], @"Can't inject with incorrect frame class.");
```
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: why is this here? it's typechecked, right?

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 18]
```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
   return self;
 }
 
+- (void)injectScopeHierarchy:(_CKChildScopeFrameRet)oldFrame
+{
+  Class componentClass = oldFrame.cls;
```
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: probably unnecessary; the compiler will do this for you in C++ if you're doing it for efficiency reasons. if you just prefer it for readability, carry on

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm [hunk 19]
```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
   return self;
 }
 
+- (void)injectScopeHierarchy:(_CKChildScopeFrameRet)oldFrame
```
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: pass const reference:

  - (void)injectChildScopeHierarchy:(const CKChildScopeHierarchy &)hierarchy;

**Comment by adamjernst** [comment]
Perfect, let's clean it up and land it.

File: ComponentKit/Core/Scope/CKComponentScopeFrame.mm (hunk 19 above)
Comment (darknoon) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: Passing const reference instead of by value is actually less efficient since it can't move in. Do you have a specific reason for this request?

Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeFrame.mm]: Nope, carry on then

File: ComponentKit/Core/CKComponentMemoizer.mm (hunk 15 above)
Comment (darknoon) [inline ComponentKit/Core/CKComponentMemoizer.mm]: Good question. I wrote some tests, so going to try changing it and re-running them. One thing is that we could probably leave the old component in the queue if it's not usable if I move the call into `dequeue`.
